servers/rendering/renderer_rd/shaders/canvas.glsl

```diff
@@ -449,7 +457,54 @@ void main() {
 
 #endif
 
-	color *= texture(sampler2D(color_texture, texture_sampler), uv);
+#ifndef USE_PRIMITIVE
+	if (bool(draw_data.flags & FLAGS_USE_MSDF)) {
+		float px_range = draw_data.ninepatch_margins.x;
+		float outline_thickness = draw_data.ninepatch_margins.y;
+		//float reserved1 = draw_data.ninepatch_margins.z;
+		//float reserved2 = draw_data.ninepatch_margins.w;
+
+		if (outline_thickness > 0) {
+			vec2 s_start = draw_data.src_rect.xy;
+			vec2 s_end = draw_data.src_rect.xy + abs(draw_data.src_rect.zw);
+			vec2 r_start = draw_data.src_rect.xy + abs(draw_data.src_rect.zw) / 3.0;
+			vec2 r_end = draw_data.src_rect.xy + abs(draw_data.src_rect.zw) * 2.0 / 3.0;
+
+			vec2 uv_s = clamp(uv, r_start, r_end);
+			uv_s = msdf_map(uv_s, r_start, r_end, s_start, s_end);
+
+			float a = 0.0;
+			float r2 = outline_thickness * outline_thickness / 4;
+			for (int y = 0; y < int(outline_thickness); y++) {
```

Comment:
With an sdf can't outline be done in a single lookup as well?   I can't remember the details of msdf off the top of my head, but I recall doing outline in a single lookup in the past.  Also should be able to get perfect antialiasing on the outline with fwidth the same as without an outline.

Reply:
Indeed, supporting soft outlines would be useful for things like soft shadows. You can already do those using shaders, but a built-in way will be more convenient and possibly faster.

Reply:
> With an sdf can't outline be done in a single lookup as well?

It is possible, but work well only for thin outlines. Maybe I'm missing something. And `FTStroker` produce self-intersections and sharp angles, causing msdfgen generation to break.

A better outline shader would be appreciated, if you have any ideas, give it a try.

> With an sdf, it would even be nice to add a 'falloff' uniform as well as thickness... Quite useful for some effects.

Existing shader is reusing existing nine patch uniform for `pixel range` and `outline thickness` which have two unused members, should be possible to add it without any extra cost.

Reply:
> It is possible, but work well only for thin outlines. Maybe I'm missing something. And `FTStroker` produce self-intersections and sharp angles, causing msdfgen generation to break.
> 
> A better outline shader would be appreciated, if you have any ideas, give it a try.

See last two posted screens for a thick SDF outline.  Once you have your geometry defined as a distance field you don't need to muck around with freetype stroker at all (shouldn't be used at all).  SDF will be more efficient and is pretty performant for doing this stuff.  Single lookup in fragment shader.